### PR TITLE
Support JWT in `clickhouse-client`

### LINKED ISSUE #17270: Implement JWT for ClickHouse
**Use case**
We have multiple end users making queries to CH through API server. API server is connected to CH using its own user. We would like to send end user credentials per query to CH, so CH can verify that the user is indeed allowed to access the database. We currently do this validation in the API server, but it would be more robust if CH validated it again.

**Describe the solution you'd like**
I don't have much experience with CH yet, but thought of one possible solution: 
1. send JWT using query settings
2. CH will decode JWT and validate that the current database specified in the context is allowed.

This might be a very narrow use case, but I would to hear thoughts on usefulness of this feature or other ways to do this.

### PR DESCRIPTION
**Pull request by thevar1able**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support JWT authentication in `clickhouse-client`

Closes https://github.com/ClickHouse/ClickHouse/issues/17270.

JWT authentication will be available only in ClickHouse Cloud.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit a5d1ac96f90dcfd3daf44e25355c283600deb9d0 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62829/a5d1ac96f90dcfd3daf44e25355c283600deb9d0/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Client/ConnectionParameters.h [hunk 1]
```diff
@@ -30,7 +31,7 @@ struct ConnectionParameters
     ConnectionParameters(const Poco::Util::AbstractConfiguration & config, std::string host);
     ConnectionParameters(const Poco::Util::AbstractConfiguration & config, std::string host, std::optional<UInt16> port);
 
-    static UInt16 getPortFromConfig(const Poco::Util::AbstractConfiguration & config, std::string connection_host);
+    static UInt16 getPortFromConfig(const Poco::Util::AbstractConfiguration & config, String connection_host);
```
Comment (tavplubix) [inline src/Client/ConnectionParameters.h:34]: Do we need to copy it? (use `const String &` or `std::move`)

File: src/Client/ConnectionPool.h [hunk 2]
```diff
@@ -123,7 +123,7 @@ class ConnectionPool : public IConnectionPool, private PoolBase<Connection>
     {
         return std::make_shared<Connection>(
             host, port,
-            default_database, user, password, SSHKey(), quota_key,
+            default_database, user, password, SSHKey(), "", quota_key,
```
Comment (tavplubix) [inline src/Client/ConnectionPool.h:126]:
```suggestion
            default_database, user, password, SSHKey(), /*jwt*/ "", quota_key,
```

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -5675,7 +5675,7 @@ std::optional<QueryPipeline> StorageReplicatedMergeTree::distributedWriteFromClu
         {
             auto connection = std::make_shared<Connection>(
                 node.host_name, node.port, query_context->getGlobalContext()->getCurrentDatabase(),
-                node.user, node.password, SSHKey(), node.quota_key, node.cluster, node.cluster_secret,
+                node.user, node.password, SSHKey(), "", node.quota_key, node.cluster, node.cluster_secret,
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:5678]:
```suggestion
                node.user, node.password, SSHKey(), /*jwt*/"", node.quota_key, node.cluster, node.cluster_secret,
```

File: src/Core/Protocol.h [hunk 4]
```diff
@@ -62,6 +62,7 @@ const char USER_INTERSERVER_MARKER[] = " INTERSERVER SECRET ";
 
 /// Marker for SSH-keys-based authentication (passed as the user name)
 const char SSH_KEY_AUTHENTICAION_MARKER[] = " SSH KEY AUTHENTICATION ";
+const char JWT_AUTHENTICAION_MARKER[] = " JWT AUTHENTICATION ";
```
Comment (tavplubix) [inline src/Core/Protocol.h:65]: Please grep all usages of USER_INTERSERVER_MARKER and SSH_KEY_AUTHENTICAION_MARKER (you will find [this place](https://github.com/ClickHouse/ClickHouse/blob/9d8f643f5b306ff02ed8e55dd776afb04e67de49/src/Access/User.cpp#L25-L37))

File: src/Client/ClientBase.cpp [hunk 5]
```diff
@@ -2015,6 +2019,15 @@ void ClientBase::processParsedSingleQuery(const String & full_query, const Strin
             connection->setDefaultDatabase(new_database);
         }
     }
+    else
```
[the remark this answers is not in the document]
Comment (thevar1able) [reply src/Client/ClientBase.cpp:2022]: `server_exception` is set elsewhere, it is class property AFAIR. Agreed on `if` styling.

File: src/Client/ConnectionParameters.h (hunk 1 above)
Comment (thevar1able) [reply src/Client/ConnectionParameters.h:34]: Can do, but I didn't even change the signature, just used a type alias.

File: src/Client/Connection.cpp [hunk 6]
```diff
@@ -335,6 +337,11 @@ void Connection::sendHello()
         performHandshakeForSSHAuth();
     }
 #endif
+    else if (!jwt.empty())
+    {
+        writeStringBinary(EncodedUserInfo::JWT_AUTHENTICAION_MARKER, *out);
+        writeStringBinary(jwt, *out);
```
Comment (vitlibar) [inline src/Client/Connection.cpp:343]: Here `jwt` is sent but I couldn't find in this PR where it's received and processed. Is this PR finished?

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 7]
```diff
@@ -5680,7 +5680,7 @@ std::optional<QueryPipeline> StorageReplicatedMergeTree::distributedWriteFromClu
         {
             auto connection = std::make_shared<Connection>(
                 node.host_name, node.port, query_context->getGlobalContext()->getCurrentDatabase(),
-                node.user, node.password, SSHKey(), node.quota_key, node.cluster, node.cluster_secret,
+                node.user, node.password, SSHKey(), /*jwt*/"", node.quota_key, node.cluster, node.cluster_secret,
```
Comment (vitlibar) [inline src/Storages/StorageReplicatedMergeTree.cpp:5683]: "" ?

Comment (thevar1able) [reply src/Storages/StorageReplicatedMergeTree.cpp:5683]: It is not used here.

File: src/Client/Connection.cpp (hunk 6 above)
Comment (thevar1able) [reply src/Client/Connection.cpp:343]: This PR is only for handling JWT in client.

Comment (vitlibar) [reply src/Client/Connection.cpp:343]: But what is the point of sending jwt from clickhouse-client if clickhouse-server can't handle it?

**Comment by clickgapai** [comment]
Hi @thevar1able — while reviewing this PR I found the following:

- **Coverage gap:** `User::setName` reserved-prefix check for ` JWT AUTHENTICATION ` marker is untested — https://github.com/ClickHouse/ClickHouse/pull/104088

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**